aten/src/ATen/cuda/CUDAGraph.cpp

```diff
@@ -0,0 +1,135 @@
+              "This CUDAGraph instance already owns a captured graph. "
+              "To capture a new graph, create a new instance.");
+
+  // For now, a CUDAGraph instance only accommodates the default generator on the device that's
+  // current when capture begins. If any op in the captured region uses a non-default generator,
+  // or a generator on another device, the offending generator will throw an error.
+  // These restrictions simplify CUDAGraph, but could be relaxed in the future:
+  // in principle, the underlying Cuda calls do permit cross-device ops to be captured.
+  auto* gen = get_generator_or_default<CUDAGeneratorImpl>(
+      c10::nullopt, cuda::detail::getDefaultCUDAGenerator());
+
+  auto options = TensorOptions().device(at::kCUDA).dtype(at::kLong);
+  offset_extragraph_ = at::empty({1}, options);
+
+  gen->capture_prologue(offset_extragraph_.data_ptr<int64_t>());
+
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  TORCH_CHECK(stream != at::cuda::getDefaultCUDAStream(),
+              "CUDA graphs must be captured on a non-default stream. "
+              "(However, after capture, it's ok to replay them on the "
+              "default stream.)");
+
+  capture_stream_ = stream;
+
+  // cudaStreamCaptureModeGlobal is the most conservative option to
+  // prevent potentially unsafe CUDA API calls during capture.  See
+  // https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g9d0535d93a214cbf126835257b16ba85
+  AT_CUDA_CHECK(cudaStreamBeginCapture(capture_stream_, cudaStreamCaptureModeGlobal));
+
+  // Stashes the current graph's uuid.
+  cudaStreamCaptureStatus status;
+  AT_CUDA_CHECK(cudaStreamGetCaptureInfo(stream, &status, &id_));
+  TORCH_INTERNAL_ASSERT(status == cudaStreamCaptureStatus::cudaStreamCaptureStatusActive);
+#else
+  TORCH_CHECK(false, "CUDA graphs may only be used in Pytorch built with CUDA >= 11.0");
+#endif
+}
+
+void CUDAGraph::capture_end() {
+#if CUDA_VERSION >= 11000
+  auto stream = at::cuda::getCurrentCUDAStream();
+
+  TORCH_CHECK(stream == capture_stream_,
+              "Capture must end on the same stream it began on.");
+  AT_CUDA_CHECK(cudaStreamEndCapture(capture_stream_, &graph_));
+  TORCH_CHECK(graph_ != NULL, "Invalid capture.");
+  has_graph_ = true;
+
+  // Trailing NULL, NULL, 0 arguments were recommended by Cuda driver people,
+  // who prefer not to report error message through these arguments moving forward
+  // (they prefer return value, or errors on api calls internal to the capture)
+  AT_CUDA_CHECK(cudaGraphInstantiate(&graph_exec_, graph_, NULL, NULL, 0));
+  has_graph_exec_ = true;
+
+  auto* gen = get_generator_or_default<CUDAGeneratorImpl>(
+      c10::nullopt, cuda::detail::getDefaultCUDAGenerator());
+  wholegraph_increment_ = gen->capture_epilogue();
+
+  // Now that we've instantiated graph_ into graph_exec_,
+  // we don't need graph_ anymore.
+  AT_CUDA_CHECK(cudaGraphDestroy(graph_));
```

Comment:
It's also handled in the CUDAGraph destructor, but I don't like that (and neither does the compiler) because it can throw an exception.

Local RAII guard here is a good idea that lets me take cudaGraphDestroy out of the destructor, but cudaGraphExecDestroy is a different story.  I don't see how to clean up graph_exec_ without cudaGraphExecDestroy in [either the destructor or a __del__ method in Python](https://github.com/pytorch/pytorch/pull/48875/files#diff-d7302d133bb5e0890fc94de9aeea4d9d442555a3b40772c9db10edb5cf36a35cR120-R134), and neither is a great option.

Reply:
Uh oh, using an RAII guard for graph_ means I call `AT_CUDA_CHECK(cudaGraphDestroy(graph_));` in the guard destructor.  I can't win here :joy: :gun:

Reply:
Cuda 11_1 build refused to compile with exceptions in destructors.  Moved graph cleanup to a reset() method called by `__del__` in a thin Python wrapper (`__del__` may or may not be a lesser evil than a throwing destructor, but it's the evil CI allows).

Reply:
Yeah, just ignore the error code, we do this in a few other places too

Reply:
? Do you mean I should go back to cleanup in the destructor, rather than `__del__`?  If so, how do I tell the CI build (I believe it was `pytorch_libtorch_linux_xenial_cuda11_1_cudnn8_py3_gcc7_build`, possibly others) not to error on the throwing destructor?  Or should I suppress the exception with a try catch?  I don't want to do the latter:  silently ignoring errors is worse than catastrophically failing.

Reply:
Yeah, other places where we are ignoring the error happen during final cleanup when the process exits (cublas handles and the like), so who cares, whereas graph desctructor should be called during runtime more or less regularly, so I agree ignoring the code is not ideal. I vote for `__del__` unless there's a better solution.

Reply:
per slack discussion, moved cleanup back to C++ destructor, but with warnings instead of throwing checks.